ooni/nettests/blocking/web_connectivity.py

```diff
@@ -0,0 +1,512 @@
+        ['backend', 'b', None, 'The web_consistency backend test helper'],
+    ]
+
+
+def is_public_ipv4_address(address):
+    try:
+        ip_address = IPv4Address(address)
+        if not any([ip_address.is_private,
+                    ip_address.is_loopback]):
+            return True
+        return False
+    except AddressValueError:
+        return None
+
+class WebConnectivityTest(httpt.HTTPTest, dnst.DNSTest):
+    """
+    Web connectivity
+    """
+    name = "Web connectivity"
+    description = ("Performs a HTTP GET request over Tor and one over the "
+                  "local network and compares the two results.")
+    author = "Arturo Filastò"
+    version = "0.1.0"
+
+    contentDecoders = [('gzip', GzipDecoder)]
+
+    usageOptions = UsageOptions
+
+    inputFile = [
+        'file', 'f', None, 'List of URLS to perform GET requests to'
+    ]
+
+    requiredTestHelpers = {
+        'backend': 'web-connectivity',
+        'dns-discovery': 'dns-discovery'
+    }
+    requiredOptions = ['backend', 'dns-discovery']
+    requiresRoot = False
+    requiresTor = False
+    followRedirects = True
+
+    # Factor used to determine HTTP blockpage detection
+    # the factor 0.7 comes from http://www3.cs.stonybrook.edu/~phillipa/papers/JLFG14.pdf
+    factor = 0.7
+    resolverIp = None
+
+    @classmethod
+    @defer.inlineCallbacks
+    def setUpClass(cls):
+        try:
+            answers = yield client.lookupAddress(
+                cls.localOptions['dns-discovery']
+            )
+            assert len(answers) > 0
+            assert len(answers[0]) > 0
+            cls.resolverIp = answers[0][0].payload.dottedQuad()
+        except Exception as exc:
+            log.exception(exc)
+            log.err("Failed to lookup the resolver IP address")
+
+
+    def inputProcessor(self, filename):
+        """
+        This is a specialised inputProcessor that also supports taking as
+        input a csv file.
+        """
+        def csv_generator(fh):
+            for row in csv.reader(fh):
+                yield row[0]
+
+        def simple_file_generator(fh):
+            for line in fh:
+                l = line.strip()
+                # Skip empty lines
+                if not l:
+                    continue
+                # Skip comment lines
+                elif l.startswith('#'):
+                    continue
+                yield l
+
+        try:
+            fh = open(filename)
+            line = fh.readline()
+            # Detect the line of the citizenlab input file
+            if line.startswith("url,"):
+                generator = csv_generator(fh)
+            else:
+                fh.seek(0)
+                generator = simple_file_generator(fh)
+            for i in generator:
+                if not i.startswith("http"):
+                    i = "http://{}/".format(i)
+                yield i
+        finally:
+            fh.close()
```

Comment:
Technically you are correct. It will be destroyed as part of the `__del__` call as soon as the reference count for it goes to zero. However if an exception is thrown before it goes out of scope, then a reference to it will be kept as part of the stack trace and hence it will not be closed. So it's more robust to generally explicitly call close on it.